### DISCUSSION TITLE
Let's design Cpp2's _requires-expression_ as a pit of success!

### DISCUSSION THREAD

**Discussion by JohelEGP** [hsutter/cppfront#588] [Suggestions]
# Let's design Cpp2's _requires-expression_ as a pit of success!

As it turns out, with Cpp1's _requires-expression_, slight changes in a _requirement_'s syntax can still result in a valid _requirement_.
This results in seemingly working `concept` declarations that actually have unintended effects.
For examples of pitfalls, I suggest you (re)read the article <https://quuxplusone.github.io/blog/2021/06/09/another-concepts-chest-mimic/> and (re)watch the linked CppCon talk.

With Cpp2, we have the opportunity to design a _requires-expression_ for Cpp2 that is a pitfall of success.

Please, have one top-level comment per (re)design.
Feel free to have top-level comments about other concerns.

### Grammars

The sky is the limit; the grammar is free real state (as Herb's willing to hand it out).
Do not think that Cpp2's _requires-expression_ has to be a single top-level grammar.
For example, each _requirement_ could be usable on its own, without being within a _requires-expression_
(so it could be an alternative of grammars, rather than the _requirement-body_ being a sequence of *requirement*s).

### Integration

Consider the greater picture of Cpp2; there might be tradeoffs to be had.

For example, in an _is-as-expression_, where the rhs argument can be a type or expression,
you can disambiguate an expression from a type by adding parentheses around it
(Herb has stated his intention to do the same for template arguments at <https://github.com/hsutter/cppfront/issues/531#issuecomment-1622252733>).

If we were to translate Cpp1's *requirement*s as-is, we could use the same formulation.
For example, instead of translating Cpp1 *requirement*s
```C++
𝘌;
typename 𝘛;
```
to Cpp2's terms
```Cpp2
𝘌;
type 𝘛;
```
they could be translated to
```Cpp2
(𝘌);
𝘛;
```

### Pitfalls

Squint at what the sources of the pitfalls of Cpp1's _requires-expression_ might be.
Besides slight changes in the syntax resulting in another valid meaning, there are also meaningless combinations.

Let's consider the first source of pitfalls, slight changes in syntax.
If we were to translate Cpp1's _compound-requirement_ as-is, we might end up with Cpp1
```C++
{ x } -> std::integral;
```
being the equivalent of Cpp2's
```Cpp2
{ x } is std::integral;
```
Unfortunately, this still runs into the pitfall that a user might instead write:
```Cpp2
x is std::integral;
```
That is a _simple-requirement_,  if we similarly translated that from Cpp1.
It changes the meaning; currently, an `is` operation always valid ([or bugged](https://github.com/hsutter/cppfront/pull/529/files#diff-9b9f6d5ded63da9a82c7f23fde117cd494a355abad3624569e71a7bf73f974deR81-R82)) (which I argue against with #492).

Herb has also stated his intention to support terse concept syntax like `<T: type is std::regular>` at <https://github.com/hsutter/cppfront/issues/575#issuecomment-1676199109>
(I'm supposing we will be able to omit `type`, and similarly declare NTTPs like `<V: _ is std::regular>`).
So this also ties back to [integration](#integration).

The other source of pitfalls is meaningless combinations.
Because a _requires-expression_ can test arbitrary expression, it can also test *requires-expression*s.
It's a meaningless combination because the SFINAE-level validity of all constructs is already being checked within a _requires-expression_.
So adding a nested `requires` doesn't add value, and not rejecting that is a source of pitfalls.
This source is enhanced by the fact that a _nested-requirement_ starts with `requires` in Cpp1.

**Comment by AbhinavK00** [comment]
>Because a requires-expression can test arbitrary expression, it can also test requires-expressions.
It's a meaningless combination because the SFINAE-level validity of all constructs is already being checked within a requires-expression.
So adding a nested requires doesn't add value, and not rejecting that is a source of pitfalls.

So if I understand correctly, you are saying that a requires expression should not be inside another requires expression, right? Because, requires clauses inside requires expressions are useful. Also, I think cpp20 already disallows that.

Also, maybe we could have different keywords for requires expressions and requires clauses because they are confusing currently.

**Reply by JohelEGP** [reply]
> Also, I think cpp20 already disallows that.

Definitely not, because it's valid and it's a pitfall.
This is what I'm talking about, from the article:
```C++
template<class T>
concept HasAButNotB = requires (T t) {
    t.a();
    !requires { t.b(); };
};
```

But maybe I'm wrong about banning `requires` within `requires`.
How else can you require that `T` doesn't have a `.b()`?

**Reply by JohelEGP** [reply]
I got an idea.

We still want to prohibit nested `requires` expressions.
What we can do is,
in addition of _nested-requirement_,
we could add its negation by adding an optional `!` at the start of a _nested-requirement_.

We use this pattern a lot in `mp-units`' unit tests to test that our generic functions don't work for unexpected inputs.
For example, see <https://github.com/mpusz/mp-units/blob/caab43cd256e4c617f57b896afb5780d1353f119/test/unit_test/static/quantity_test.cpp#L51-L59>.
So by
- prohibiting `requires !requires { 𝘴𝘵𝘶𝘧𝘧; }`,
- permitting `!requires { 𝘴𝘵𝘶𝘧𝘧; }`, and
- prohibiting nested *requires-expression*s,

we remove the pitfall and keep the valid use cases.

**Reply by AbhinavK00** [reply]
One concern
```cpp
!requires { 𝘴𝘵𝘶𝘧𝘧; }
```
is always be satisfied in cpp1 so we should prohibit that one. Or maybe I misunderstood and you actually mean allowing it in cpp2 which would lower to cpp1's
```cpp
requires !requires { 𝘴𝘵𝘶𝘧𝘧; }
```
Right?

**Reply by JohelEGP** [reply]
That's right (the latter).

**Comment by AbhinavK00** [comment]
I'll give some code examples
The following is your example which is valid but wrong
```cpp
template<class T>
concept HasAButNotB = requires (T t) {
    t.a();
    !requires { t.b(); };
};
```

The following example shows why I think "it" is banned in cpp20, this will error out
```cpp
template<class T>
concept HasAButNotB = requires (T t) {
    t.a();
    requires { t.b(); };
};
```

The following is the correct way to express our requirement
```cpp
template<class T>
concept HasAButNotB = requires (T t) {
    t.a();
    requires !requires { t.b(); };
};
```
Also, I think you're right about banning requires inside requires upto some extent. requires _**expressions**_ should not be allowed (probably aren't in cpp20 but have to check that) but requires _**clauses**_ are still useful as demonstrated by above example.

This also supports my point about having high distinction between requires _**clauses**_ and requires _**expressions**_ maybe even upto having different keywords for them. And cppfront definitely needs to simplify this design space largely so users don't fall into pitfalls like above.

**Reply by JohelEGP** [reply]
What you call a requires clause in the example is actually a third grammar that begins with `requires`.
It's a _nested-requirement_ (the other two being _requires-clause_ and _requires-expression_).

> This also supports my point about having high distinction between requires _**clauses**_ and requires _**expressions**_ maybe even upto having different keywords for them. And cppfront definitely needs to simplify this design space largely so users don't fall into pitfalls like above.

I'm really glad you voiced out this concern.
It's been a known issue forever.
But Cpp1 just wouldn't dare to take up more keywords,
and you could even say the contextual overloading of `requires` makes sense.

I considered this, but can't come up with any solution.
And if I ever came across a better suggestion that would fit Cpp2, I certainly don't remember.
Maybe you can find a suggestion out there whose frame really fits Cpp2.

**Reply by AbhinavK00** [reply]
Isn't that same as a requires clause which is nested though? it works the same, takes a boolean and makes acts a toggle on/off switch in the encompassing requires expression.
Sry if it's a stupid question,  I just woke up

**Reply by JohelEGP** [reply]
The grammar of a _requires-clause_ is more constrained.
Which I believe would be a semantic constraint in Cpp2,
which just uses the same grammar of _nested-requirement_, namely _logical-or-expression_.

**Comment by JohelEGP** [comment]
Here's the CppCon talk's pitfalls for *requires-expression*s.
- <https://youtu.be/CXn02MPkn8Y?t=2337>:
  _simple-requirement_.
  Cpp1: `-t -> T` vs. `{ -t } -> T`.
  Cpp2: `x is T` vs. `{ x } is T` (my literal translation from the Cpp1 grammar).
- <https://youtu.be/CXn02MPkn8Y?t=2418>:
  boolean _simple-requirement_, e.g. `sizeof(T) == 4;`.
  Maybe lower a Cpp2 _simple-requirement_ `𝘌;`
  to a Cpp1 _compound-requirement_ `{ 𝘌 } -> cpp2::not_`[_`boolean-testable`_](https://eel.is/c++draft/concept.booleantestable#concept:boolean-testable)`_or_static_assert;`.
  The error message should point out that maybe it was meant to be a constraint.
  And that if it isn't the case, to convert it the _compound-requirement_ `{ 𝘌 } is bool;` or to discard it (`_ = 𝘌;`).
- <https://youtu.be/CXn02MPkn8Y?t=2455>:
  Cpp1: `noexcept ( ++t );`, `requires noexcept( ++t );`, `{ ++t } noexcept;`.
  Cpp2: N/A (ATM?).

**Reply by AbhinavK00** [reply]
The first one is naturally solved in cpp2 because there's no operator `->` and because we're using better behaved `is`.
Like, let the users write 
```cpp
x is T
```
and we know for sure they mean 
```cpp
{ x } -> T
``` 
And if they really mean `x ->T` then in cpp2, it would be written as `x*.T`.

**Reply by JohelEGP** [reply]
This isn't about `->` in Cpp2.
Rather, I'm suggesting that a _compound-requirement_ in Cpp1 `{ 𝘌 } -> 𝘛` translates to Cpp2 `{ 𝘌 } is 𝘛`,
and so the same pitfall of the _simple-requirement_ in Cpp1 `𝘌 -> 𝘛` also translates to Cpp1 `𝘌 is 𝘛`.

**Reply by AbhinavK00** [reply]
In cpp1, `E -> T` acts as if you're dereferencing a pointer E and accessing it's member T which is unintuitive but this problem is naturally solved in cpp2 because we're using `is`. So this pitfall need not to be taken care as cpp2 solves it naturally.

**Reply by JohelEGP** [reply]
I do not think it would be solved in Cpp2 as-is.
As I see it, these two formulations are valid and mean different things:
```Cpp2
𝘌 is 𝘛;
{ 𝘌 } is 𝘛;
```
I argue that the `is` operator shouldn't always be valid at #492.
So I need the first syntax to not be equivalent to the second (or equivalent to Cpp1 `{ 𝘌 } -> 𝘛`)
because `𝘌 is 𝘛` is just a normal Cpp2 expression and I want it to be testable in a _simple-requirement_.

**Reply by AbhinavK00** [reply]
I think I understand now. 
```cpp
E is T
```
This checks for correctness of the expression i.e. if the query is well-formed or not while
```cpp
{ E } is T
```
Checks if the query is actually true or not.

In cpp1 terms, the second one would be expressed as 
```
requires E is T
```
Therefore this falls under 

> Semantic rule: Not on compound-requirement, but on a expression requirement.
An expression does not satisfy boolean-testable (i.e., if it does, the implementation shall emit a diagnostic).

And therefore should be changed in accordance to that.

**Reply by JohelEGP** [reply]
I mean for `{ 𝘌 } is 𝘛;` to lower to `{ 𝘌 } -> 𝘛;`.

**Reply by JohelEGP** [reply]
Now, I suppose `𝘛` would have to be a concept for this to be an issue.
Cppfront's Cpp2 doesn't support the expression `0 is std::integral` right now,
so the validity of `𝘌 -> 𝘛` and `{ 𝘌 } -> 𝘛` is mutually exclusive.
The first requires `𝘛` to be a type, and the second `𝘛` to be a `concept`.
I suspect we want `0 is std::integral` to work anyways, but we just don't have the means to make it work.

**Reply by AbhinavK00** [reply]
> I mean for { 𝘌 } is 𝘛; to lower to { 𝘌 } -> 𝘛;.

Ok, sorry for the confusion. I get it now but `E is T` shouldn't lower `E -> T` which you already implied.

**Reply by JohelEGP** [reply]
I didn't imply that.
I did mention that the problem of missing the braces in Cpp1 applies to Cpp2.
So Cpp1 `𝘌 -> 𝘛` could be a typo for `{ 𝘌 } -> 𝘛`,
and Cpp2 `𝘌 is 𝘛` could be a typo for `{ 𝘌 } is 𝘛`.

**Reply by AbhinavK00** [reply]
Yes and the following semantic rule proposed by you should take care of this

> Semantic rule: Not on compound-requirement, but on a expression requirement.
An expression does not satisfy [boolean-testable](https://eel.is/c++draft/concept.booleantestable#concept:boolean-testable) (i.e., if it does, the implementation shall emit a diagnostic).
Intent: To prevent https://github.com/hsutter/cppfront/discussions/588#discussioncomment-6715051 bullets 1 and 2.
Those pitfalls become compile-time errors, and the user should do one of these: 
\
Move 𝘌; outside the requires-expression if that's what they meant
(i.e., as a term of a requires-clause when 𝘌 is a constant expression that satisfies [boolean-testable](https://eel.is/c++draft/concept.booleantestable#concept:boolean-testable)).
Discard it with _ = 𝘌; if it's actually expressed correctly and 𝘌 can satisfy [boolean-testable](https://eel.is/c++draft/concept.booleantestable#concept:boolean-testable).
Transform it to { 𝘌 } is std::convertible_to<bool>; or something along those lines if that's what they meant.

And therefore `E is T` should be invalid and the correct way of writing that would be `{ E is T } is bool`.

**Reply by JohelEGP** [reply]
Yes, that's part of my proposed design: <https://github.com/hsutter/cppfront/discussions/588#discussioncomment-6715505>.

**Reply by JohelEGP** [reply]
The second bullet, also at ["What's New in Compiler Explorer? - Matt Godbolt - C++ on Sea 2023"](https://youtu.be/O5sEug_iaf4?t=2333).

**Comment by JohelEGP** [comment]
Here's my design, based on "no gratuitous difference from Cpp1".
It's basically a compilation of my other comments and a restriction on `inspect`.
If my statements about them are correct, they could even be proposals for Cpp1.

```C++
// requires-expression:
//     'requires' parameter-declaration-list? requirement-body
//
// requirement:
//     type-id ';'          // Cpp1 type-requirement
//     expression ';'       // Cpp1 simple-requirement
//     compound-requirement
//     nested-requirement
```

Semantic rule: A (possibly negated) nested _requires-expression_ may only appear as a term of a _nested-requirement_.
Intent: To prevent the subject of <https://quuxplusone.github.io/blog/2021/06/09/another-concepts-chest-mimic/>.

```C++
// compound-requirement:
//     { expression } 'throws'? is-type-constraint? ';'
```

Maybe not really `is-type-constraint`, but whatever grammar Herb adds or uses for `<T: type is std::regular>` (see <https://github.com/hsutter/cppfront/issues/575#issuecomment-1676199109>).
Most *compound-requirement*s will have to be `throws`, just like how most function types would have to be `throws` with my #526.

Semantic rule: Not on _compound-requirement_, but on a _expression_ _requirement_.
An _expression_ does not satisfy [_`boolean-testable`_][] (i.e., if it does, the implementation shall emit a diagnostic).
Intent: To prevent <https://github.com/hsutter/cppfront/discussions/588#discussioncomment-6715051> bullets 1 and 2.
Those pitfalls become compile-time errors, and the user should do one of these:
- Move `𝘌;` outside the _requires-expression_ if that's what they meant
  (i.e., as a term of a _requires-clause_ when `𝘌` is a constant expression that satisfies [_`boolean-testable`_][]).
- Discard it with `_ = 𝘌;` if it's actually expressed correctly and `𝘌` can satisfy [_`boolean-testable`_][].
- Transform it to `{ 𝘌 } is std::convertible_to<bool>;` or something along those lines if that's what they meant.

[_`boolean-testable`_]: https://eel.is/c++draft/concept.booleantestable#concept:boolean-testable

```C++
// nested-requirement:
//     '!'? 'requires' logical-or-expression ';'
```

When `!` is present, results in the negated value of the same _nested-requirement_ without `!`.
Intent: To help prevent the subject of <https://quuxplusone.github.io/blog/2021/06/09/another-concepts-chest-mimic/>
by making the formulation of a "has no" requirement a first-class construct.

If proposing these to Cpp1, `requires !requires { 𝘴𝘵𝘶𝘧𝘧; }` can remain valid, but possibly deprecated.

**Reply by JohelEGP** [reply]
I forgot to mention `inspect`.
I think it should be banned, just like nested *requires-expression*s.
An `inspect` does a lot of SFINAE internally, so allowing it is likely to become a source of pitfalls.
Instead, users should break off from an `inspect` what it is that they actually want to constrain.
And this is true for all proposals, regardless of syntax and semantics (so equally for Cpp1 and Cpp2).

**Reply by JohelEGP** [reply]
I'm implementing this at <https://github.com/JohelEGP/cppfront/tree/requires-expression>.
I had to split the negative _nested-requirement_
into its own grammar _negative-requirement_:
```C++
    //G nested-requirement:
    //G     'requires' logical-or-expression ';'
    //G negative-requirement:
    //G     '!' 'requires' requirement
```
It's a bit worrying that, in source code, they can visually differ by the `!` only.
I think that adding the restriction on booleans (see above) to the _requirement_ of the _negative-requirement_
will be enough to not introduce any new source of pitfalls.

**Reply by JohelEGP** [reply]
> Semantic rule: Not on _compound-requirement_, but on a _expression_ _requirement_.
> An _expression_ does not satisfy [_`boolean-testable`_][] (i.e., if it does, the implementation shall emit a diagnostic).
> Intent: To prevent <https://github.com/hsutter/cppfront/discussions/588#discussioncomment-6715051> bullets 1 and 2.
> Those pitfalls become compile-time errors, and the user should do one of these:
> - Move `𝘌;` outside the _requires-expression_ if that's what they meant
>   (i.e., as a term of a _requires-clause_ when `𝘌` is a constant expression that satisfies [_`boolean-testable`_][]).
> - Discard it with `_ = 𝘌;` if it's actually expressed correctly and `𝘌` can satisfy [_`boolean-testable`_][].
> - Transform it to `{ 𝘌 } is std::convertible_to<bool>;` or something along those lines if that's what they meant.

I wonder if this is still necessary when an expression that beings with an identifier has to be parenthesized.
The `sizeof(T) == 4;` from the talk (<https://github.com/hsutter/cppfront/discussions/588#discussioncomment-6715051>) is still affected in Yoda's style (`4 == sizeof(T)`).

**Reply by JohelEGP** [reply]
I need a better translation for `{ 𝘌 } -> 𝘊`.
My currently proposed `{ 𝘌 } throws is 𝘊` reads worse.

This still reads fine: `{ 𝘌 } noexcept -> 𝘊` to `{ 𝘌 } is 𝘊`.

**Reply by JohelEGP** [reply]
The solution to the above is to use a comma: `{ 𝘌 } throws, is 𝘊;`.
This talk might be of interest: [David Sankel :: The Semicolon is a Lie](https://www.youtube.com/watch?v=ICf_6L1kJcE).
Particularly relevant is this clip of 6 s: <https://youtube.com/clip/UgkxAIFQJNcIlMqIzRzPw_pwnONnVOxVMGT6>.

**Reply by AbhinavK00** [reply]
One thing I wanted to say, maybe you can consider using `->` like cpp instead of is. Yes, I know it's consistent with rest of the langauge but it also has one disadvantage of introducing an expression, like this one https://github.com/hsutter/cppfront/discussions/588#discussioncomment-6716479.
The advantage of `->` is that it does not introduce an expression. `E -> T` is naturally ill-formed in cpp2.
This is not a suggestion but merely an option you could consider.

**Reply by JohelEGP** [reply]
That might be possible.
We can even do it with `throws`.
```Cpp2
{ 𝘌 } throws, is 𝘊;
𝘌 throws -> 𝘊; // No `{}` anymore.
𝘌 -> 𝘊; // Non potentially-throwing 𝘌.
```
Unfortunately, this isn't only inconsistent with the planned introduction of `is 𝘊` by <https://github.com/hsutter/cppfront/issues/575#issuecomment-1676199109>,
but it's also inconsistent with Cpp1.
I'll stick with being more consistent with Cpp1 _and_ consistent with Cpp2.

**Reply by JohelEGP** [reply]
Why did I have to do this change: <https://github.com/JohelEGP/cppfront/commit/17caed18579bc40a1f03c5dcab94dc8535ec3a25#diff-e7cad51d49cc330296fa31a67e79b5d1497bd068bafaabd30bca2951f0f6ab19L4-R5>?
Somehow, it was making the tests I expected to pass, not pass.
It turns out I was using a concept, which is ["always"](https://wg21.link/LWG2290) valid.
So the negative requirement was to never pass.
What I wanted to test was indirection, or `std::iter_reference_t<T>`.
Here, the semantic requirement of non-convertibility to boolean would have helped me.
EDIT: Actually, that currently parses like a type requirement. So my result above raced with moving _negative-requirement_ into its own grammar with commit 17caed18579bc40a1f03c5dcab94dc8535ec3a25.

**Reply by JohelEGP** [reply]
> Semantic rule: A (possibly negated) nested _requires-expression_ may only appear as a term of a _nested-requirement_.
> Intent: To prevent the subject of <https://quuxplusone.github.io/blog/2021/06/09/another-concepts-chest-mimic/>.

I think this is completely taken care of by
- requiring an expression requirement to not be convertible to bool, and
- adding negative requirement as a first-class construct.

**Reply by JohelEGP** [reply]
> Most *compound-requirement*s will have to be `throws`, just like how most function types would have to be `throws` with my #526.

All would-be *simple-requirement*s in Cpp1 have to be Cpp2 *compound-requirement*s with `throws` specified.

**Reply by JohelEGP** [reply]
> I'm implementing this at <https://github.com/JohelEGP/cppfront/tree/requires-expression>.

Opened #596 for this.
